app/src/main/java/io/github/jbellis/brokk/SessionManager.java

```diff
@@ -370,10 +371,49 @@ public void saveHistory(ContextHistory ch, UUID sessionId) {
         sessionExecutorByKey.submit(sessionId.toString(), () -> {
             try {
                 Path sessionHistoryPath = getSessionHistoryPath(sessionId);
+
+                // Snapshot current tasklist.json (if present) before we rewrite the zip
+                String taskListJsonSnapshot = null;
+                if (Files.exists(sessionHistoryPath)) {
+                    try (var fs = FileSystems.newFileSystem(sessionHistoryPath, Map.of())) {
+                        Path tlPath = fs.getPath("tasklist.json");
+                        if (Files.exists(tlPath)) {
+                            taskListJsonSnapshot = Files.readString(tlPath);
+                        }
+                    } catch (IOException ioe) {
+                        logger.debug(
+                                "Could not snapshot existing tasklist.json for session {}: {}",
+                                sessionId,
+                                ioe.getMessage());
+                    }
+                }
```

Comment:
I'd extract the common pieces regarding reading and writing task list into separate methods, and reuse here and in `read/writeTaskList` methods.